arrows/vxl/convert_image.cxx

```diff
@@ -275,21 +245,62 @@ class convert_image::priv
   {
   }
 
+  // Convert a fraction of images to gray
+  template < typename Type >
+  vil_image_view< Type >
+  random_gray_conversion( vil_image_view< Type > const& src,
+                          double const random_fraction );
+
   // Apply appropriate transforms
-  template < typename ipix_t > vil_image_view< ipix_t >
+  template < typename ipix_t >
+  vil_image_view< ipix_t >
   apply_transforms( vil_image_view_base_sptr& view );
 
   // Scale and convert the image
-  template < typename opix_t, typename ipix_t > vil_image_view< opix_t >
-  scale_and_convert( vil_image_view< ipix_t > input );
+  template < typename opix_t, typename ipix_t >
+  vil_image_view< opix_t >
+  scale_and_convert( vil_image_view< ipix_t >& input );
 
   std::string format;
   bool single_channel;
   double scale_factor;
   double random_grayscale;
   double percentile_norm;
+
+  std::random_device random_device;
+  std::mt19937 engine{ random_device() };
+  std::uniform_real_distribution< double > dist{ 0.0, 1.0 };
 };
 
+// ----------------------------------------------------------------------------
+template < typename Type >
+vil_image_view< Type >
+convert_image::priv
+::random_gray_conversion( vil_image_view< Type > const& src,
+                          double const random_fraction )
+{
+  auto random_value = dist( engine );
+
+  if( random_value < random_fraction )
```

Comment:
...not necessary, but at this point you almost might as well write:
```suggestion
  if( dist( engine ) < random_fraction )
```

(On that note, however, you might also consider renaming `dist` → `random_dist`, and `engine` → `random_engine`, or maybe `rng`.)